spikeextractors/extractors/neuroscopeextractors/neuroscopeextractors.py

```diff
@@ -0,0 +1,451 @@
+import os
+
+from os import listdir
+from os.path import isfile, join
+
+try:
+    import bs4
+    import lxml
+    HAVE_BS4_LXML = True
+except ImportError:
+    HAVE_BS4_LXML = False
+
+class NeuroscopeRecordingExtractor(BinDatRecordingExtractor):
+    
+    """
+    Extracts raw neural recordings from large binary .dat files in the neuroscope format.
+    
+    The recording extractor always returns channel IDs starting from 0.
+    
+    The recording data will always be returned in the shape of (num_channels,num_frames).
+
+    Parameters
+    ----------
+    folder_path : str
+        Path to the folder containing the .dat file.
+    """
+    extractor_name = 'NeuroscopeRecordingExtractor'
+    installed = True  # check at class level if installed or not
+    is_writable = True
+    mode = 'folder'
+    installation_mesg = 'please install bs4 and lxml to use this extractor'  # error message when not installed
+            
+    def __init__(self, folder_path, subset_channels=None):
+        assert HAVE_BS4_LXML, self.installation_mesg
+        RecordingExtractor.__init__(self)
+        self._recording_file = folder_path
+        
+        fpath_base, fname = os.path.split(folder_path)
+        xml_filepath = os.path.join(folder_path, fname + '.xml')
+        dat_filepath = os.path.join(folder_path, fname + '.dat')
+        
+        with open(xml_filepath, 'r') as xml_file:
+            contents = xml_file.read()
+            soup = BeautifulSoup(contents, 'lxml')
+            # Normally, this would be a .xml, but there were strange issues
+            # in the write_recording method that require it to be a .lxml instead
+            # which also requires all capital letters to be removed from the tag names
+        
+        n_bits = int(soup.nbits.string)
+        dtype='int'+str(n_bits)
+        num_channels = int(soup.nchannels.string)
+        
+        if subset_channels is not None:
+            num_channels = len(subset_channels)
+        
+        sampling_frequency = float(soup.samplingrate.string)
+        
+        BinDatRecordingExtractor.__init__(self, dat_filepath, sampling_frequency=sampling_frequency,
+                                          dtype=dtype, numchan=num_channels)
+        
+        self._kwargs = {'folder_path': str(Path(folder_path).absolute()),
+                        'subset_channels': subset_channels}
+        
+        
+    @staticmethod
+    def write_recording(recording, save_path, dtype='int32'):
+        """ Convert and save the recording extractor to Neuroscope format
+
+        parameters
+        ----------
+        recording: RecordingExtractor
+            The recording extractor to be converted and saved
+        save_path: str
+            Full path to desired target folder
+        dtype: str
+            Data type to be used in writing. Will throw a warning if stored recording type from get_traces() does not match.
+        """
+        _, RECORDING_NAME = os.path.split(save_path)
```

Comment:
would this fail if one passes a relative path? e.g. `save_path='recording'`

I recommend looking into `pathlib` for handling paths, but let's keep it for later. I can quickly modify it once it's merged

Reply:
Yes, relative paths could definitely cause issues if the passed path was not in correspondence to the working directory at the time of the call. That said, we only ever pass absolute paths in the functions that call these extractors, so I'll just add that requirement for now in the docstring.

For our purposes, I can't see needing the ability to do a better job on relative paths since we never use these in a scripting manner that would tend to rely more heavily on working directory location; but if you think there is potential for other users that may want to use the functions in such a way, I leave that to your discretion.

Reply:
More importantly, I suppose, is that the RECORDING_NAME may not have to be exactly the same as the last element of the save_path, though it simply happens to be in the datasets we are prototyping on. I'll try to generalize that a bit more as well.

Reply:
Actually, never mind. It's pretty easy to just coerce the absolute path and extract the specific name rather than expecting adherence to any particular format or directory. Fixing this now.

Reply:
I've tried to generalize this as much as possible. The paths that are passed within the test_extractor.py need a particular format to work with the absolute conversions, but theoretically, a user could pass a really sparse relative path and it _should_ work.